ASoC: SOF: Intel: Revisit IMR boot sequence

ujfalusi opened the pull request:
The sequence for IMR boot is essentially the same as normal boot with the
difference that instead of DMA from host the firmware is loaded from IMR.

Re-structure the code to use the existing sequence and also add fallback
handling in case the IMR boot fails.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

kv2019i commented:
FYI @jairaj-arava

ujfalusi commented:
Suspend/resume looks clean all around the board if not even better than in other PR tests around the time (which might be just random fluctuation).

ranj063 commented:
> Suspend/resume looks clean all around the board if not even better than in other PR tests around the time (which might be just random fluctuation).

@ujfalusi should this have a fixes tag?

ujfalusi commented:
@ranj063, I'm not sure about the fixes tag, the original imr implementation worked as far as I can recall, but it was skipping some of the steps.
@keyonjie, what is your opinion?

The IMR support was added by this commit:
5fb5f5118512 ("ASoC: SOF: Intel: hda-loader: add IMR restore support")

ranj063 commented:
> @ranj063, I'm not sure about the fixes tag, the original imr implementation worked as far as I can recall, but it was skipping some of the steps. @keyonjie, what is your opinion?
> 
> The IMR support was added by this commit: [5fb5f51](https://github.com/thesofproject/linux/commit/5fb5f51185126059e1d7eb4e452e08c7b17c3301) ("ASoC: SOF: Intel: hda-loader: add IMR restore support")

I still think we should add the fixes tag. If we're fixing the sequence, i feel like thats the right thing to do

ujfalusi commented:
@ranj063, right. This patch basically drops the original patch content and the implementation is new, but it is fair to give @keyonjie credit for his work on this.

I can update the commit message with instructions and proposed commit message for upstreaming

plbossart commented:
@ranj063 the only benefit of a Fixes: tag is that it will be picked up by Linux stable.

I don't think we have any reports of failures today, do we? I would think this doesn't really qualify as a must-have for -stable. 

I also don't think we need to give credit for past work, this was done already in the initial contribution, this is a new effort by @ujfalusi alone.

ujfalusi commented:
Changes since v1:
- converted to fixup, but we need two patch to fix up when sending upstream...
- introduced HDA flag: imrboot_supported and set it up after first firmware boot.

ujfalusi commented:
> @ranj063 the only benefit of a Fixes: tag is that it will be picked up by Linux stable.
> 
> I don't think we have any reports of failures today, do we? I would think this doesn't really qualify as a must-have for -stable.

I don't think we have failure either, but I think it is beneficial to follow official sequence, which is exactly the same as normal boot with two differences:
- the ROM control message should not have the PURGE bit set (but still should be sent)
- waiting for FW_ENTERED instead of INIT status from the DSP

We also have the benefit of recovering from failed IMR boot, if it ever happens.
 
> I also don't think we need to give credit for past work, this was done already in the initial contribution, this is a new effort by @ujfalusi alone.

OK, I thought a `fixes!` patch was requested. I normally use fixes tag for patches fixing broken things, but the initial IMR sequence worked so far. Having said that we can opt to add the tag when sending upstream, let's see then.

plbossart commented:
This still seems stuck @greg-intel, but Jenkins indicates the job #7490 completed two days ago. Will restart the test.

plbossart commented:
SOFCI TEST

plbossart commented:
We missed the window for daily tests, @ujfalusi can you please fix the indentation issue reported by checkpatch.

````
WARNING: suspect code indent for conditional statements (16, 32)
#161: FILE: sound/soc/sof/intel/hda-loader.c:531:
+		if (!sof_debug_check_flag(SOF_DBG_IGNORE_D3_PERSISTENT) &&
[...]
+				hdev->imrboot_supported = true;

WARNING: suspect code indent for conditional statements (16, 32)
#208: FILE: sound/soc/sof/intel/icl.c:68:
+		if (!sof_debug_check_flag(SOF_DBG_IGNORE_D3_PERSISTENT) &&
[...]
+				hdev->imrboot_supported = true;

````

let's merge tomorrow.

greg-intel commented:
@plbossart looks like the SOFCI command worked for this PR?  The issues with the Jenkins checks not getting updated has been fixed, so things should be more-or-less smooth now.

plbossart commented:
> @plbossart looks like the SOFCI command worked for this PR? The issues with the Jenkins checks not getting updated has been fixed, so things should be more-or-less smooth now.

yes, I had to restart the tests manually to override the previous issue, but it seems like we are back to normal.

plbossart commented:
> let's merge tomorrow.

And of course my PR #3566 introduced a conflict, sorry about that @ujfalusi, you'll have to rebase as well.

ujfalusi commented:
Changes since v2:
- fixed the indentation issues reported by checkpatch
- rebased on sof-dev to resolve the conflict

ujfalusi commented:
> > let's merge tomorrow.
> 
> And of course my PR #3566 introduced a conflict, sorry about that @ujfalusi, you'll have to rebase as well.

No issue!

plbossart commented:
@ranj063 can you approve+merge?
